include/ofi_util.h

```diff
@@ -1156,7 +1157,54 @@ enum {
 	OFI_OPT_TCP_FI_ADDR = -FI_PROV_SPECIFIC_TCP
 };
 
+struct util_queue {
+	struct dlist_entry list;
+	dlist_func_t *match_func;
+};
+
+struct util_rx_entry {
+	struct fi_peer_rx_entry	peer_entry;
+	uint64_t		ignore;
+	int			multi_recv_ref;
+};
+
+struct util_srx_ctx {
+	struct fid_peer_srx	peer_srx;
+	struct util_queue	recv_queue;
+	struct util_queue	trecv_queue;
+	bool			dir_recv;
+	size_t			min_multi_recv_size;
+	uint64_t		rx_op_flags;
+	uint64_t		rx_msg_flags;
+	size_t			iov_limit;
+
+	struct util_cq		*cq;
+	struct util_queue	unexp_msg_queue;
+	struct util_queue	unexp_tagged_queue;
+	struct ofi_bufpool	*rx_pool;
+	ofi_spin_t		lock;
+};
```

Comment:
Since this is meant to be generic, here's a comparison with tcp/net:
```
struct xnet_srx {
	struct fid_ep		rx_fid;
	struct xnet_domain	*domain;
	struct slist		rx_queue;
	struct slist		tag_queue;
	struct ofi_dyn_arr	src_tag_queues;
	struct ofi_dyn_arr	saved_msgs;

	struct xnet_xfer_entry	*(*match_tag_rx)(struct xnet_srx *srx,
						 struct xnet_ep *ep,
						 uint64_t tag);

	uint64_t		tag_seq_no;
	struct ofi_bufpool	*buf_pool;
	uint64_t		op_flags;
	size_t			min_multi_recv_size;

	/* Internal use when srx is part of rdm endpoint */
	struct xnet_rdm		*rdm;
	struct xnet_cq		*cq;
	struct util_cntr	*cntr;
};
```
It's worth noting that tcp needs to support srx with msg ep's.  The tcp msg queue (rx_queue) is based on slist not dlist.  dlist insertion/removals are expensive vs slist.  (dlist is useful when you need to remove an item from the middle of a list *and* you're not traversing the list at the time.)  For tagged msgs, it maintains index-based (src addr) queues (src_tagged_queues, and saved_msgs).  There isn't an rx_pool associated with the srx.

No additional locking is used by srx, as a progress level lock is held instead.  That lock is a mutex (genlock actually, as it can be a no-op), since the calls made while holding it can block.  Even the spinlock above is held during malloc calls, which can also block.  We need to figure out how locking works across providers.  In any case, if we tried to use this with tcp, an srx level lock doesn't quite work.  Internally, the correct lock is already held.  Only an external caller would need to acquire the right lock.

The unexpected message queue ends up being an issue when messages are received prior to the peer having been inserted into the local AV.  I don't know how this code handles that situation.

Reply:
The idea of this implementation (for now) is to have a simple, non-optimized srx. My goal is to change the dlist to an slist and add the src address unexpected messaging optimization that tcp/net has. But for now, this implementation matches what shm uses which is why I've only updated shm to use this implementation. But the goal is improve this generic implementation to support other providers without a loss of optimizations/support.

I think we definitely need to figure out locking across providers. There will have to be some sort of resource control, obviously. Perhaps we could use the new atomic queue for a lockless approach? The limitation there would be that the pool would not be dynamic, but if we stick to a specific rx_size then probably would be ok? (I haven't thought that out at all, just spit balling.)

As for dealing with unexpected messages before the address is inserted into the AV, this implementation does not take that into account at all so I'll definitely need to add that. How does tcp fix that?

Reply:
TCP is selective about which messages it wants to save.  If the receiving endpoint doesn't have the peer in the local AV, it leaves the message unprocessed until later.  Basically, the srx can't save just any unexpected message that an endpoint wants to toss at it.  There's no source address available, so it doesn't know where in the src_tag_queues array to go.  Storing it somewhere else means that once the peer is inserted into the AV, any messages that have been received by that peer already need their source address updated.  (I think this problem shows up with the mpich spawn tests.)  We may need to require this behavior for users of peer srx queues.

Based on tcp, for a util_srx, the lock and rx_pool aren't general requirements.  The buf_pool in the xnet_srx is a lie.  It's not used anywhere.

The other wonderful discovery I recently made is that MPI will sometimes post a receive for a message *after* the peer process has exited.  shm probably doesn't care, because it doesn't support msg ep's, but verbs would.  In the case of tcp and verbs, the local msg ep would be disconnected and likely cleaned up prior to the receive being claimed by the app.  This is why you see both cq and cntr fields in the tcp srx struct.  I need to be able to complete a saved receive message after the receiving endpoint is no more.

Reply:
I'd like to compare the tcp, verbs, rxm, and efa (if there's one) implementations, plus thinking about the locking more.  One of the other implementations or some combination may be a better starting point.

With FI_THREAD_DOMAIN, we probably don't need any locking at all.  The peer providers will also be sharing a domain.  For anything else, we'll need serialization, but need to agree on what that will look like.

Reply:
Are the peer providers always supposed to be sharing a domain? That's not how I'm doing it....

Reply:
They don't have to be, but if they are, they have more information available, like the threading model used by the app.  Specifically, pairing shm with another provider, peer domain sharing makes sense.  Conceptually, that's what the app would view.

Can this help with locking?  I don't know yet.

Reply:
EFA creates unexp queue per ep and per peer both. The former case is for unspecified addr receive and the other one is for directed receive. 

I would like to see the util recv queue did the same thing, but it may need some more work. Currently my plan is to use EFA's own queue for SRX because the API is not specifying the queue implementation. Finally we could help making this code in util to remove the code duplication.

Reply:
Does EFA have a base SRX implementation somewhere?

Reply:
The SRX is not implemented in EFA provider yet. We are working on it, hopefully will be available within a month or so.

Reply:
I could find 2 relevant implementations for srx in tree, shm and tcp, and they basically have different implementations.  I'm not convinced that a util_srx will be all that usable.  I don't see that it hurts anything to have it, but it may not be ideal.

This really targets receiving on an rdm ep.  For example, the verbs srx targets msg eps's, and is even more different than shm or tcp because that's mapping to hardware.  It doesn't have tag matching or need to support complex protocols.

We may need to clarify that this is only usable with a single AV or single rdm ep + the peer.